lean4.html.markdown

```diff
@@ -0,0 +1,587 @@
+-/
+def collatz_next (n : Nat) : Nat :=
+  if n % 2 = 0 then n / 2 else 3 * n + 1
+
+def iter (k : Nat) (f: Nat → Nat) :=
+  match k with
+  | Nat.zero => fun x => x
+  | Nat.succ k' => fun x => f (iter k' f x)
+
+theorem collatz : ∀ n, n > 0 → ∃ k, iter k collatz_next n = 1 := sorry
+
+/-
+Now, some "corner cases" in logic.
+-/
+
+/-
+The proposition `True` is something that can be trivially proved.
+
+`True.intro` is a constructor for proving `True`. Notice that it needs no
+inputs.
+-/
+theorem obvious : True := True.intro
+
+/-
+On the other hand, there is no constructor for `False`.
+
+We have to use `sorry`.
+-/
+theorem impossible : False := sorry
+
+/-
+Any `False` in the hypothesis allows us to conclude anything.
+
+Written in term style, we use the eliminator `False.elim`. It takes a proof of
+`False`, here `h`, and concludes whatever is the goal.
+-/
+theorem nonsense (h : False) : 0 = 1 := False.elim h
+
+/-
+The `contradiction` tactic uses any `False` in the hypothesis to conclude the
+goal.
+-/
+theorem more_nonsense (h : False) : 1 = 2 := by contradiction
+
+/-
+To illustrate constructive vs classical logic, we now prove the contrapositive
+theorem.
+
+The forward direction does not require classical logic.
+-/
+theorem contrapositive_forward' : (p → q) → (¬q → ¬p) :=
+  fun pq => fun hqf => fun hp => hqf (pq hp)
+/-
+Use the definition `¬q := q → False`. Notice that we have to construct `p →
+False` given `p → q` and `q → False`. This is just function composition.
+-/
+
+/-
+The above proof, using tactics.
+-/
+theorem contrapositive_forward : (p → q) → (¬q → ¬p) := by
+  intro hpq
+  intro
+  intro hp
+  specialize hpq hp
+  contradiction
+
+/-
+The reverse requires classical logic.
+
+Here, we are required to construct a `q` given values of following types:
+
+  - `(q → False) → (p → False)`.
+  - `p`.
+
+This is impossible without using the law of excluded middle.
+-/
+theorem contrapositive_reverse' : (¬q → ¬p) → (p → q) :=
+  fun hnqnp =>
+  Classical.byCases
+    (fun hq  => fun  _ => hq)
+    (fun hnq => fun hp => absurd hp (hnqnp hnq))
+/-
+Law of excluded middle tells us that we will have a `q` or a `q → False`. In the
+first case, it is trivial to construct a `q`, we already have it. In the second
+case, we give the `q → False` to obtain a `p → False`.  Then, we use the fact
+(in constructive logic) that given `p` and `p → False`, we can construct
+`False`. Once, we have `False`, we can construct anything, and specifically `q`.
+-/
+
+/-
+Same proof, using tactics.
+-/
+theorem contrapositive_reverse : (¬q → ¬p) → (p → q) := by
+  intro hnqnp
+  intro hp
+  have emq := Classical.em q
+  cases emq
+  case inl _ => assumption
+  case inr h => specialize hnqnp h ; contradiction
+
+/-
+To illustrate how we can define an work with axiomatic systems. Here is a
+definition of Groups and some proofs directly translated from "Topics in
+Algebra" by Herstein, Second edition.
+-/
+
+/-
+A class defines a set of axioms
```

Comment:
This is misleading.
`class A where a : Int` is perfectly valid. (Yes, this can be viewed as an axiom, just don't do it.)